src/model_theory/terms_and_formulas.lean

```diff
@@ -362,6 +379,137 @@ begin
+| rel {l : ℕ} {R : L.relations l} {ts : fin l → L.term (α ⊕ fin n)} :
+    is_atomic (R.bounded_formula ts)
+
+lemma is_atomic.induction {P : L.bounded_formula α n → Prop} {φ : L.bounded_formula α n}
+  (h : is_atomic φ)
+  (he : ∀ (t₁ t₂ : L.term (α ⊕ fin n)), P (bd_equal t₁ t₂))
+  (hr : ∀ {l : ℕ} (R : L.relations l) (ts : fin l → L.term (α ⊕ fin n)), P (R.bounded_formula ts)) :
+  P φ :=
+begin
+  induction h,
+  { apply he },
+  { apply hr }
+end
+
+lemma is_atomic_bd_equal (t₁ t₂ : L.term (α ⊕ fin n)) :
+  is_atomic (t₁.bd_equal t₂ : L.bounded_formula α n) :=
+is_atomic.equal
+
+lemma is_atomic_rel {m} (R : L.relations m) (ts) :
+  is_atomic (R.bounded_formula ts : L.bounded_formula α n) :=
+is_atomic.rel
+
+lemma is_atomic.relabel {m : ℕ} {φ : L.bounded_formula α m} (h : φ.is_atomic)
+  (f : α → β ⊕ (fin n)) :
+  (φ.relabel f).is_atomic :=
+begin
+  refine h.induction _ _;
+  intros,
+  { exact is_atomic_bd_equal _ _ },
+  { exact is_atomic_rel _ _ }
+end
+
+/-- A quantifier-free formula is a formula defined without quantifiers. These are all equivalent
+to boolean combinations of atomic formulas. -/
+inductive is_qf : L.bounded_formula α n → Prop
+| falsum : is_qf falsum
+| of_is_atomic {φ} (h : is_atomic φ) : is_qf φ
+| imp {φ₁ φ₂} (h₁ : is_qf φ₁) (h₂ : is_qf φ₂) : is_qf (φ₁.imp φ₂)
+
+lemma is_qf.induction {P : L.bounded_formula α n → Prop} {φ : L.bounded_formula α n}
+  (h : is_qf φ)
+  (hf : P (⊥ : L.bounded_formula α n))
+  (ha : ∀ {ψ : L.bounded_formula α n}, is_atomic ψ → P ψ)
+  (himp : ∀ {φ₁ φ₂} (h₁ : P φ₁) (h₂ : P φ₂), P (φ₁.imp φ₂)) :
+  P φ :=
+begin
+  induction h with _ ih0 _ _ _ _ ih1 ih2,
+  { exact hf },
+  { exact ha ih0 },
+  { exact himp ih1 ih2 }
+end
+
+lemma is_atomic.is_qf {φ : L.bounded_formula α n} : is_atomic φ → is_qf φ :=
+is_qf.of_is_atomic
+
+lemma is_qf_bot : is_qf (⊥ : L.bounded_formula α n) :=
+is_qf.falsum
+
+lemma is_qf.not {φ : L.bounded_formula α n} (h : is_qf φ) :
+  is_qf φ.not :=
+h.imp is_qf_bot
+
+lemma is_qf.relabel {m : ℕ} {φ : L.bounded_formula α m} (h : φ.is_qf)
+  (f : α → β ⊕ (fin n)) :
+  (φ.relabel f).is_qf :=
+begin
+  refine h.induction _ _ _,
+  { exact is_qf_bot },
+  { intros _ h,
+    exact (h.relabel f).is_qf },
+  { intros _ _ h1 h2,
+    exact h1.imp h2 }
+end
+
+/-- Indicates that a bounded formula is in prenex normal form - that is, it consists of quantifiers
+  applied to a quantifier-free formula. -/
+inductive is_prenex : ∀ {n}, L.bounded_formula α n → Prop
+| of_is_qf {n} {φ : L.bounded_formula α n} (h : is_qf φ) : is_prenex φ
+| all {n} {φ : L.bounded_formula α (n + 1)} (h : is_prenex φ) : is_prenex φ.all
+| ex {n} {φ : L.bounded_formula α (n + 1)} (h : is_prenex φ) : is_prenex φ.ex
+
+lemma is_qf.is_prenex {φ : L.bounded_formula α n} : is_qf φ → is_prenex φ :=
+is_prenex.of_is_qf
+
+lemma is_atomic.is_prenex {φ : L.bounded_formula α n} (h : is_atomic φ) : is_prenex φ :=
+h.is_qf.is_prenex
+
+lemma is_prenex.induction {P : ∀ {n}, L.bounded_formula α n → Prop}
```

Comment:
^

Reply:
As with the other induction lemmas, there appears to be something about `rec_on` that doesn't work right for all these term-mode proofs, but would work if I did it in tactic mode with `exact`.